### ISSUE TITLE
History and Open with last file not working

### ISSUE THREAD

**Issue by KenMaltby** [koreader/koreader#2006]
* KOReader version: 379>
* Device:Kobo AuraHD

#### Issue
Preference for SDR may have lost us the ability to start back on the page we were reading from. 
#### Steps to reproduce
Suspend (in the hope of the data being saved, with new code), unsuspend.
Return to File Manager (again to try and insure that the position is recorded). Exit KOReader.
Reboot, when KOReader starts it opens in the File Manager, even though the Open in last file option is checked.
Using the book's entry in History opens the book several chapters before the point you were at.

**Comment by KenMaltby** [comment]
The loss of these features is really annoying, I had not realized how much I was depending on them. The open in last file, on the page I was reading last, especially.  The History entry may be opening on the last page saved under the old history process.

**Comment by AlanSP1** [comment]
Yes, this is vital for reading any kind of book. Hope it will be solved soon.

**Comment by houqp** [comment]
Yes, I will take a look at it later tonight hopefully. I have been moving this week so was not able to get any work done :(

**Comment by Frenzie** [comment]
Welcome to the new place!

**Comment by houqp** [comment]
OK, my room is still a huge mess right now and it's bed time already so definitely not today. Let's push it to tomorrow :P

**Comment by Frenzie** [comment]
So I was looking at this and I noticed something odd.

I tested with this EPUB
http://www.gutenberg.org/ebooks/17489

Go to page 2 → opens on page 1
Go to page 12 → opens on page 1
Go to page 13 → opens on page 13
Go to page 19 → opens on page 1

The problem seems to be in the `last_xpointer`.

For page 12:
```
	["last_xpointer"] = "/body/DocFragment[1]/body.0",
	[…]
	["percent_finished"] = 0.012257405515832,
```

For page 13:
```
	["last_xpointer"] = "/body/DocFragment[1]/body/p[3]/text().678",
	[…]
	["percent_finished"] = 0.013278855975485,
```

So for all I know the problem might be in crengine. For instance introduced in https://github.com/koreader/crengine/commit/159bcb3f20e461a41c204d338ad2a004a59487f2#diff-77f6525e33d2fd1aeeb073a64dd8d759 or https://github.com/koreader/crengine/commit/f120e2810e47e058e9006ca373175a54d239cebe#diff-77f6525e33d2fd1aeeb073a64dd8d759

That's it for now. =P

**Comment by Frenzie** [comment]
@KenMaltby Can you confirm my suspicion that this problem shouldn't occur on anything handled by MuPDF (e.g. PDF, DjVu)?

**Comment by KenMaltby** [comment]
It is looking the same with "first aid fm4_25x11.pdf"  -  "Start with last opened file" still opens on the File Manager page that lists the file, instead of in the book, on the page I was reading, as was the old way.  From History, it opens to the page it was originally on when I last looked at it a month ago, where it opened at the beginning of this test, not the much later page I was on for the test.

I haven't tried with a totally new book, that has never been opened before under the old code.

**Comment by Frenzie** [comment]
I haven't been able to reproduce this with PDF. Is that document available online?

**Comment by KenMaltby** [comment]
Not sure this is where I got it but it is here: https://archive.org/details/FM4-25x11.  I might have used the OPDS catalog.

**Comment by KenMaltby** [comment]
I just had a thought, I have KSM set with the option: "deleteSDRsOnboard=true".  I will retest with it false.

**Comment by Frenzie** [comment]
Yeah, I can't seem to reproduce that within the first 20 pages or so (as well as some random going to chapters and browsing) with scrollmode both on and off. It simply reloads where I left off.

> I just had a thought, I have KSM set with the option: "deleteSDRsOnboard=true". I will retest with it false.

That doesn't occur anywhere in the current source according to a quick search, so it shouldn't affect anything either.

**Comment by KenMaltby** [comment]
Wierdness,  with the KSM option false the History function works sometimes, but not always.  The Open in last file only works if I don't reboot in KSM, or PowerOff and On.  Otherwise; it opens in the File Manager.

**Comment by KenMaltby** [comment]
@Frenzie  Could you relook at your fix for #1986

**Comment by Frenzie** [comment]
It's @chrox' but I will if I find some time. ;)

**Comment by houqp** [comment]
Not able to reproduce on my side. @KenMaltby what is SDR? Are saying everything works fine (including open last file and progress restore) if the device is not rebooted?

**Comment by houqp** [comment]
The issue mentioned by @Frenzie seems to be a separate bug. I have created a new ticket for it.

**Comment by KenMaltby** [comment]
Now, if I use the "Home" icon in the reader menu and exit KOReader back to the KSM menu, then select "Start koreader", I get the Info message box with it opening the file and it returns to the page I was on.  ( That used to be what also happened if you were to use the KSM PowerOff function or rebooted.)  If you use the Home icon and then the KSM PowerOff function, when you restart the device and KSM Auto selects KOReader it opens in the File Manager, (no message box, no book).  

[Oddly, if I then exit the File Manager back to KSM and select "Start koreader" it then opens properly.
This may provide a work around that gets me back to where I was in the book.] 

 If I use the PowerOff then restart and select "Start koreader" before the autoselect occurs, it still opens in the File Manager. 

{OT- KUDOs for the hierarchical Table of Contents, this issue has me using it a lot.}

As I understand it; the SDR folders are now "preferred" over the History entries for storing the file/book related info developed while reading. (The History entries and the .lua files in the SDR folders look very similar, the SDR .lua files contain some additional data.)

**Comment by djalma2k** [comment]
> Steps to reproduce
> 
> Suspend (in the hope of the data being saved, with new code), unsuspend.
> Return to File Manager (again to try and insure that the position is recorded). Exit KOReader.
> Reboot, when KOReader starts it opens in the File Manager, even though the Open in last file option is checked.
> Using the book's entry in History opens the book several chapters before the point you were at.

I already tried to reproduce it using version 2015-11-392 but I can't. Everything about the OP problem (@KenMaltby) works as it should in my Glo HD... (I haven't any options added to KSM to modify the koreader's behavior. Only set the preference to open koreader by default)
Maybe I don't understand the problem properly... anyway, related to this:

> opens the book several chapters before the point you were at.

Well something similar happens from version (>2015-11-31X) but it happens because the images are processed or rendered in a different way that in previous versions then all the book pages are consequently modified.
Regards.

**Comment by houqp** [comment]
> Now, if I use the "Home" icon in the reader menu and exit KOReader back to the KSM menu, then select "Start koreader", I get the Info message box with it opening the file and it returns to the page I was on. ( That used to be what also happened if you were to use the KSM PowerOff function or rebooted.) If you use the Home icon and then the KSM PowerOff function, when you restart the device and KSM Auto selects KOReader it opens in the File Manager, (no message box, no book).

Is there a feature in KSM that auto selects KOReader after a restart? By no message box, no book, you mean an empty blank screen? Or a filemanager window with empty file list?

After a reboot, can you check to see if there is a valid `lastfile` entry in `settings.reader.lua` before starting KOReader? The line that loads the last file is at: https://github.com/koreader/koreader/blob/master/reader.lua#L158. You can also add a print there to see if it's loading anything.

> [Oddly, if I then exit the File Manager back to KSM and select "Start koreader" it then opens properly.
This may provide a work around that gets me back to where I was in the book.] If Koreader's Open in last file is unchecked, it will open in the File Manager (default page) {This is a different page from the File Manager page that you get on returning from PowerOff or Reboot.}

> As I understand it; the SDR folders are now "preferred" over the History entries for storing the file/book related info developed while reading. (The History entries and the .lua files in the SDR folders look very similar, the SDR .lua files contain some additional data.)

Ha, I see, you are talking about the SDR folders. I thought it was a KSM feature. The SDR migration should not affect open last file feature though.

**Comment by houqp** [comment]
> Well something similar happens from version (>2015-11-31X) but it happens because the images are processed or rendered in a different way that in previous versions then all the book pages are consequently modified.

@djalma2k, changes in typesetting should not affect the restore too much since we don't record the last page number, but a pointer to internal document structure like image or paragraph. So if it goes back to serval chapters before, then it's a bug. Is it still happening to you?

**Comment by KenMaltby** [comment]
@houqp; 
Yes, a filemanager window with empty file list, upon return from PowerOff or Reboot.
The last file entry looks right, but it has nothing to indicate the progress or page.
